fetchNodeModules: init

zhaofengli opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

This PR adds a new fetcher, `fetchNodeModules`, which can download deterministic dependency sets from the public NPM registry according to the NPM lock file (`package-lock.json`). This is useful for ensuring that Node.js packages are able to use dependencies at the exact versions specified by upstream, avoiding silent breakages caused by updates to the shared `nodePackages` package set which does not respect semantic versioning. The fetcher will also greatly simplify the work required to package Node.js projects with complex dependencies without sacrificing reproducibility.

###### Background

Currently, adding a package that makes use of the NPM ecosystem to Nixpkgs is done by adding all NPM dependencies it requires to one of the following places, both of which come with their own downsides:

<details>
<summary>Shared `nodePackages` package set</summary>

We keep a shared list of NPM packages in `pkgs/development/node-packages/node-packages.json`. When people add packages to this list, a script (`generate.sh`) is run which fetches the latest versions of *all* packages in the list from NPM, writing the URLs to retrieve the packages along with their hashes into `node-packages.nix`. This workflow has several major downsides:

- **Unrelated changes always get introduced.** When the generator script is run, all packages in the shared list are updated at once, regardless of whether they are necessary or not.
- **Semantic versioning is violated.** The generator always fetches the latest version of the packages (sans several manual exceptions), without taking compatibility into consideration.
- **Merge conflicts are a frequent occurrence.** PRs proposing changes to the shared package set (e.g., #126664, #115474) are often met with merge conflicts as the generator script often introduces large, unrelated changes to the shared `node-packages.nix`.
- **The package update process is slow.** Fetching all of the shared packages is often a slow process for maintainers. 

All of these inevitably lead to silent breakages of unrelated packages. An example of silent breakage can be found in a proposed addition of the BGPAlerter package (#115474). A major update to `js-yaml` (3.14.1 -> 4.0.0) was introduced in [a PR](https://github.com/NixOS/nixpkgs/pull/108596) that updates rust-analyzer package. Unfortunately, this release is not backwards-compatible (indicated by the increase of the major release number) and it removes several deprecated functions that were available in earlier versions (e.g., `safeLoad`). For BGPAlerter, the configuration wizard (run when there is no configuration) works without issue while actually starting the monitoring daemon will fail because of the removed `safeLoad` function. In other words, such breakages can be invisible even from those directly testing the packages.
</details>

<details>
<summary>Private dependency set</summary>

Alternatively, maintainers can also opt to separate their packages from the global shared package list. They work similarly to the global list, except at a smaller scale with changes affecting only the package itself. A tool like `node2nix` and `yarn2nix` is used to convert the NPM/Yarn dependencies (with version constraints, and in most cases, lock files) into Nix expressions containing a list of URLs and hashes corresponding to each dependency. Dependencies are downloaded according to the version constraints specified in the package's manifest (`package.json`), and in most cases, a lock file (`yarn.lock`, `package-lock.json`) will be used which specifies the exact version for each required dependency. Because of that, breakages do not occur for packages using their private dependency sets.

Due to the interconnected nature of the NPM ecosystem, the generated Nix expressions can be [extremely long](https://github.com/NixOS/nixpkgs/blob/cd3ed54f6ea1c13d45c6772b4752ae6d2ff35997/pkgs/applications/networking/cluster/spacegun/node-packages.nix), often with line counts in the thousands or even tens of thousands. As a result, reviewers can be [reluctant](https://github.com/NixOS/nixpkgs/pull/115474#discussion_r590050056) accepting new packages with their private dependency sets. The shared package set remains the preferred way for smaller leaf packages, and thus has to suffer from the issues listed above.

A list of packages that have their own dependency sets can be found by searching for `offline_cache =` and `generated by node2nix` in Nixpkgs.

</details>

###### `fetchNodeModules`

`fetchNodeModules` is similar to the `fetchCargoTarball` fetcher for Rust packages in that it relies on the native language-specific package manager (npm) to complete the entire process of resolving and downloading dependencies, resulting in a bundle of frozen packages. A fixed-output derivation is used to allow npm network access and the results are compared against a predetermined hash in the Nix expression.

The process is reproducible because like it relies on a lock file which specifies the exact version for each required dependency along with hashes created from the its contents, like `yarn2nix` and `node2nix` mentioned above. The difference is that the dependency list is not converted to Nix and only standard tools for the language are used. Reproducibility is enforced both in npm (lock file) and Nix (hash of final output).

It is important to note that the [original lock file](https://github.com/dvallin/spacegun/blob/86d42ae8dbfd5b76df1fc837b7a035c35e700b26/package-lock.json) and [the translated Nix expressions](https://github.com/NixOS/nixpkgs/blob/cd3ed54f6ea1c13d45c6772b4752ae6d2ff35997/pkgs/applications/networking/cluster/spacegun/node-packages.nix) from the aforementioned tools contain the same level of information. `fetchNodeModules` does not work with projects without a lock file.

###### Example

A tarball of all runtime dependencies of BGPAlerter can be created with the expression below:

```
pkgs.nodePackages.fetchNodeModules {
  src = fetchFromGitHub {
    owner = "nttgin";
    repo = "BGPalerter";
    rev = "v1.28.1";
    sha256 = "sha256-Y0atkJfZxjUOGPQ3goXS/YD5SsX9ZjpbM0Nc5IuaFP4=";
  };

  makeTarball = true;
  sha256 = "sha256-850tsQSYWECziz5qMoTUKRpIF7wKMuRj1e54VeWEOV4=";
}
```

###### See also

- https://github.com/NixOS/nixpkgs/pull/115474#issuecomment-870899532
- https://news.ycombinator.com/item?id=27371665

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

zhaofengli committed:
fetchNodeModules: init

zhaofengli commented:
Note that this is just a draft implementation. One thing that is puzzling me right now is that with the following expression, the output directory simply disappears after build:

```nix
pkgs.nodePackages.fetchNodeModules {
  src = pkgs.fetchFromGitHub {
    owner = "nttgin";
    repo = "BGPalerter";
    rev = "v1.28.1";
    sha256 = "sha256-Y0atkJfZxjUOGPQ3goXS/YD5SsX9ZjpbM0Nc5IuaFP4=";
  };

  makeTarball = false;
  exposeBin = true;
  sha256 = "sha256-aQziIJ8plNFhZRVEbNwb7pKgnVp6rMKlo8UR2iA5cbk=";
}
```

Would like to know if this problem is reproducible by others and what the solution is.

ajs124 commented:
What does this do that [npmlock2nix](https://github.com/tweag/npmlock2nix) (or one of the other 20 attempts to do npm things with nix) doesn't?

Mic92 commented:
> What does this do that [npmlock2nix](https://github.com/tweag/npmlock2nix) (or one of the other 20 attempts to do npm things with nix) doesn't?

I think npmlock2nix is doing import from derivation, which this tool does not.

Mic92 commented:
`fetchNodeModules` output should be tested on different platforms if it produces different results.

Mic92 commented on pkgs/build-support/fetchnodemodules/default.nix:
```diff
@@ -0,0 +1,101 @@
+        echo
+        exit 1
+    fi
+    if [[ ! -f package-lock.json ]]; then
```
I think npm also supports yarn.lock as well: https://blog.npmjs.org/post/621733939456933888/npm-v7-series-why-keep-package-lockjson.html

zhaofengli replied:
It does with the normal `npm i` which will automatically generate a `package-lock.json` from `yarn.lock`, but not the deterministic `npm ci` which we are using which refuses to perform any change to the lock files.

zhaofengli commented:
> > What does this do that [npmlock2nix](https://github.com/tweag/npmlock2nix) (or one of the other 20 attempts to do npm things with nix) doesn't?
> 
> I think npmlock2nix is doing import from derivation, which this tool does not.

Also this tool doesn't attempt to parse the `package.json` / `package-lock.json` or otherwise Nixify the dependencies at all - It simply delegates the fetching to the actual `npm`, since [`npm ci`](https://blog.npmjs.org/post/171556855892/introducing-npm-ci-for-faster-more-reliable.html) provides the reproducibility we need, which is why I compared it to `fetchCargoTarball` (used internally by `buildRustPackage`) which does the same thing but with `cargo`.

SuperSandro2000 commented on pkgs/build-support/fetchnodemodules/default.nix:
```diff
@@ -0,0 +1,101 @@
+  '';
+
+  installPhase = ''
+    if [[ ! -z $makeTarball ]]; then
```
```suggestion
    if [[ -n $makeTarball ]]; then
```

SuperSandro2000 commented on pkgs/build-support/fetchnodemodules/default.nix:
```diff
@@ -0,0 +1,101 @@
+    else
+        # Copy all modules into a directory
+        mkdir $out
+        cp -r node_modules $out
```
Could we move the files under lib, share or opt?

zhaofengli replied:
Sure, it makes sense to follow the conventions since we are having a `bin` (if `exposeBin` is set) as well. Could you also test if you can reproduce the problem that I'm seeing with `exposeBin`?

SuperSandro2000 replied:
I don't really have the time right now to fiddle with this.

SuperSandro2000 commented on pkgs/build-support/fetchnodemodules/default.nix:
```diff
@@ -0,0 +1,101 @@
+        cp -r node_modules $out
+
+        # Expose executables in node_modules/.bin
+        if [[ ! -z $exposeBin ]]; then
```
```suggestion
        if [[ -n $exposeBin ]]; then
```

SuperSandro2000 commented on pkgs/build-support/fetchnodemodules/default.nix:
```diff
@@ -0,0 +1,101 @@
+    fi
+
+    # Whether to install dev dependencies
+    if [[ ! -z $production ]]; then
```
```suggestion
    if [[ -n $production ]]; then
```

zhaofengli commented:
Pushed a fix to change the directory structure (`$out/lib/node_modules`) as well as the timestamps. The hashes in the examples have been updated.

roberth reviewed (changes requested):
> Reproducibility is enforced both in npm (lock file) and Nix (hash of final output).

Enforcement in Nix happens, but in a very narrow way. If you forget to update the hash, you will be using stale dependencies, unknowingly, until "hopefully" someone else tries to build it without your store or binary cache failing the fetcher, detecting the problem. Considering that these will end up on cache.nixos.org, these problems can linger for a long time.
It's also worth considering that these store paths will duplicate a lot of files that are shared store paths in case of other solutions. (hard link optimization only applies to a local nix store; nix-copy-closure and other forms of distribution such as docker images are still affected)

> The hashes in the examples have been updated.

We don't have a method for finding and updating these when a change logic inevitably requires this again. This is especially true when other projects call this function.

This might be alleviated somewhat by including a hash of relevant inputs in the `name`, forcing a new store path when those inputs change. However, the selection of those inputs is a trade-off between usability and correctness. Too many and you'll burden the user with unnecessary hash updates. Too few and we'll have reproducibility problems. This is not a choice we should have to make.

Specifically with npm we don't have to make this trade-off because it provides us with perfectly usable lock files that we can parse with Nix to fetch dependencies in a _reliably_ reproducible way.

Mic92 commented:
> Enforcement in Nix happens, but in a very narrow way. If you forget to update the hash, you will be using stale dependencies, unknowingly, until "hopefully" someone else tries to build it without your store or binary cache failing the fetcher, detecting the problem. Considering that these will end up on cache.nixos.org, these problems can linger for a long time.

This problem can be solved by adding the lock file to the dependencies as well and than diff the lock file from the fixed-derivation output with lock file delivered in the source.

> It's also worth considering that these store paths will duplicate a lot of files that are shared store paths in case of other solutions. (hard link optimization only applies to a local nix store; nix-copy-closure and other forms of distribution such as docker images are still affected)

Having thousands of node derivations also comes with a cost: evaluation memory and nixpkgs download/unpack size.

> 
> > The hashes in the examples have been updated.
> 
> We don't have a method for finding and updating these when a change logic inevitably requires this again. This is especially true when other projects call this function.
> 
> This might be alleviated somewhat by including a hash of relevant inputs in the `name`, forcing a new store path when those inputs change. However, the selection of those inputs is a trade-off between usability and correctness. Too many and you'll burden the user with unnecessary hash updates. Too few and we'll have reproducibility problems. This is not a choice we should have to make.
> 
> Specifically with npm we don't have to make this trade-off because it provides us with perfectly usable lock files that we can parse with Nix to fetch dependencies in a _reliably_ reproducible way.

Does not really scale in nixpkgs. Dependency trees are so large for the average nodejs package that the added file size is an not acceptable overhead to store all lock files in nixpkgs. Otherwise we would have switched tools like https://github.com/nix-community/yarn2nix or similar. We cannot use import from derivation because this would increase the evaluation time.

Mic92 commented:
To add another data sample. Peertubes yarn.lock file is 8000 lines. This is a huge amount for a single package even without the full dependency tree. I think this fetcher here is the only sane approach for nixpkgs.

roberth commented:
I expect this fetcher to be implementable as a non-fixed output recursive derivation that
1. takes a `fetchurl`ed lock file as an input
2. fetches the modules using `fetchurl` during the build
3. creates a link farm in `$out`

Recursive Nix is implemented but experimental, so while we could try it today, we can't use it by default in Nixpkgs. Eelco has raised a concern

> Probably we should also disallow instantiating/building fixed-output derivations (specifically, those that access the network[...])

However, I believe the impurity can be [addressed](https://github.com/NixOS/nix/pull/3205#issuecomment-879975406) without disallowing it. If for whatever reason that doesn't work out, this idea seems to be compatible with RFC40 ret-cont recursive Nix; a more restrictive form of recursion. Perhaps even RFC92, although I really can't confirm that now.

With this solution on the horizon, I believe we can use this PR in the interim without creating a _structural_ problem with Nix's perceived reliability.

SuperSandro2000 commented:
> takes a fetchurled lock file as an input

That wont work for everything because upstream does not always provide an (up to date) lock file.

> fetches the modules using fetchurl during the build

npm can take anything as a downloaded file and we already fetching random git repos in nodePackages right now.

roberth commented:
> > takes a fetchurled lock file as an input
> 
> That wont work for everything because upstream does not always provide an (up to date) lock file.

While true, I do not think it's relevant to my comment. It is a more general problem but the solution is easy enough: make a PR to update the lock file, and fetch from the PR. Preferably tag it in the fork as well to avoid loss on rebase, but that's yet another topic.

> > fetches the modules using fetchurl during the build
> 
> npm can take anything as a downloaded file and we already fetching random git repos in nodePackages right now.

I intended `fetchurl` as an example. Any derivation can be used, including those of `fetchgit`.

Mic92 commented:
> I expect this fetcher to be implementable as a non-fixed output recursive derivation that
> 
>     1. takes a `fetchurl`ed lock file as an input
> 
>     2. fetches the modules using `fetchurl` during the build
> 
>     3. creates a link farm in `$out`

The algorithm you are describing is effectively import-from-derivation. It is not a good idea to implement it this way because this would force a nix eval process to download all node packages just to evaluate nixpkgs. This is slow and takes a lot of disk space.
Correct me if I am wrong.

roberth commented:
> The algorithm you are describing is effectively import-from-derivation.

It is not. Evaluation for the (large number of) fetcher calls happens inside the sandbox thanks to recursive Nix. I've clarified the original comment.

> It is not a good idea to implement it this way because this would force a nix eval process to download all node packages just to evaluate nixpkgs.

To the rest of Nixpkgs, a recursive Nix `fetchNodeModules` appears and performs like any other single derivation. Thanks to recursive Nix, the expensive parts are moved into the sandbox, where it can run later, in parallel and remotely.

> This is slow and takes a lot of disk space and moves computational work from builder to evaluator (bad for how Nix CIs currently work).
> Correct me if I am wrong.

Hercules CI has solved the user experience around import-from-derivation. It doesn't have such an evaluation bottleneck and it's clear when its evaluator is waiting for IFD. IFD is still sequential as of today, but that's not a problem in practice, unless you're building at the scale of Nixpkgs.

Mic92 commented:
> To the rest of Nixpkgs, a recursive Nix `fetchNodeModules` appears and performs like any other single derivation. Thanks to recursive Nix, the expensive parts are moved into the sandbox, where it can run later, in parallel and remotely.

Ok. But it is not clear when we will be able to use this feature. So far a nix release is just pushed forward year by year by adding more experimental features. To me it seems we could add it with a checksum and when time comes we could switch over to recursive nix by dropping the need for a checksum.

> 
> Hercules CI has solved the user experience around import-from-derivation. It doesn't have such an evaluation bottleneck and it's clear when its evaluator is waiting for IFD. IFD is still sequential as of today, but that's not a problem in practice, unless you're building at the scale of Nixpkgs.

That's nice however hercules is still slower than both plain nix and hydra (which uses multiple threads).

roberth commented:
> not clear when we will be able to use this feature.

> and when time comes we could switch over to recursive nix by dropping the need for a checksum.

So we [agree](https://github.com/NixOS/nixpkgs/pull/128749#issuecomment-880240710).

> hercules evaluation is still slower than both plain nix and hydra

Thanks for letting me know. I'll start to optimize eval soon.

andir commented:
> > What does this do that [npmlock2nix](https://github.com/tweag/npmlock2nix) (or one of the other 20 attempts to do npm things with nix) doesn't?
> 
> I think npmlock2nix is doing import from derivation, which this tool does not.

Nope, it doesn't. It works in hydras restricted eval mode. You can use it with IFD if you don't want to copy the lock file into the repo.


Give that this PR proposes yet another big FOD I am really not in favor of this. We already had enough trouble in the Go and Rust world due to these issues.

I was originally planning to propose npmlock2nix for nixpkgs but I want it to be more mature first.

Mic92 commented:
> > > What does this do that [npmlock2nix](https://github.com/tweag/npmlock2nix) (or one of the other 20 attempts to do npm things with nix) doesn't?
> > 
> > 
> > I think npmlock2nix is doing import from derivation, which this tool does not.
> 
> Nope, it doesn't. It works in hydras restricted eval mode. You can use it with IFD if you don't want to copy the lock file into the repo.
> 
> Give that this PR proposes yet another big FOD I am really not in favor of this. We already had enough trouble in the Go and Rust world due to these issues.
> 
> I was originally planning to propose npmlock2nix for nixpkgs but I want it to be more mature first.

This is also not a solution just look how large a package-lock.json is: https://github.com/andig/evcc/blob/master/package-lock.json
That might work for projects that have the package-lock.json but certainly not for this one. It also still will increase evaluation time due to large dependency graph it needs to process. I don't see any other scalable solution but this fixed-input derivations and maybe recursive nix in the future.

andir commented:
> > > > What does this do that [npmlock2nix](https://github.com/tweag/npmlock2nix) (or one of the other 20 attempts to do npm things with nix) doesn't?
> > > 
> > > 
> > > I think npmlock2nix is doing import from derivation, which this tool does not.
> > 
> > Nope, it doesn't. It works in hydras restricted eval mode. You can use it with IFD if you don't want to copy the lock file into the repo.
> > 
> > Give that this PR proposes yet another big FOD I am really not in favor of this. We already had enough trouble in the Go and Rust world due to these issues.
> > 
> > I was originally planning to propose npmlock2nix for nixpkgs but I want it to be more mature first.
> 
> npmlock2nix is also not a solution just look how large a package-lock.json is: https://github.com/andig/evcc/blob/master/package-lock.json. That might work for projects that have the package-lock.json in the repo but certainly not for nixpkgs. It also still will increase evaluation time due to large dependency graph it needs to process. I don't see any other scalable solution but this fixed-input derivations and maybe recursive nix in the future.

Perhaps we shouldn't add any of those packages to Nixpkgs then. Probably super unpopular opinion but even without flakes we have great mechanism for composition.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/future-of-npm-packages-in-nixpkgs/14285/1

dotlambda commented:
> Perhaps we shouldn't add any of those packages to Nixpkgs then. Probably super unpopular opinion but even without flakes we have great mechanism for composition.

If we remove all existing ones, I 100% agree on that. The [NUR](https://github.com/nix-community/NUR) is a better place for them.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/future-of-npm-packages-in-nixpkgs/14285/2

SuperSandro2000 commented:
> Perhaps we shouldn't add any of those packages to Nixpkgs then. Probably super unpopular opinion but even without flakes we have great mechanism for composition.

And what do we do with the packages we already have that do this?

Also this doesn't solve the problem.

> We already had enough trouble in the Go and Rust world due to these issues.

Do we? The only problem I know of is when the hash generation changes when the go/rust are updated. Otherwise those packages are a lot easier to use than having to use node2nix.

roberth commented:
https://github.com/NixOS/nixpkgs/pull/136022 is a promising technique for testing fetchers.

> Enforcement in Nix happens, but in a very narrow way. If you forget to update the hash, you will be using stale dependencies, unknowingly, until "hopefully" someone else tries to build it without your store or binary cache failing the fetcher, detecting the problem.

I don't think we should use it to solve this problem though, because all these dependencies will be re-fetched and stored at distinct paths, polluting the binary cache. This would happen even when the actual output is unchanged.

It is suitable for writing a few test cases for this fetcher though, which can be run reliably via `nix-build` and ofborg without having to worry about bad cache hits.

mkg20001 commented:
To add my 2 cents: I'm currently maintaining https://github.com/mkg20001/nix-node-package which uses the lockfile directly, as in parses it at build time using builtins.fromJSON and replaces all the http paths in package-lock.json with file:// paths pointing to the nix store

This works great for smaller projects. Not sure about the usability in nixpkgs, as evaluation this way consumes a lot of resources because the lockfile has to be parsed and transformed during eval. It might be interesting with recursive nix.

The approach is quite different but it might be a good inspiration.

Mic92 commented:
We also have https://github.com/nix-community/npmlock2nix, which does the same but still lock files are quite large in both cases.

yu-re-ka commented:
BTW, the output hash differs on aarch64-darwin and x86_64-linux. When `createTarball` is set to `false`, it produces the same hash on both systems.

happysalada commented:
If I understand correctly, here is where we stand

`fetchNodeModules`:
Pros:
- could potentialy be used right away
- will help separating the dependency that nodePackages have upon each other. (at the moment in order to add or update one of the nodePackages, you need to update all of them).
- separating any package from the nodePackages cluster potentially reduces the amount of generated code today.
- would make contributing new javascript packages much easier.
Cons:
- can only be used with package-lock.json, so there might be a need to commit a lock-file, or to have upstream have their lock-file up to date.
- Uses FOD, so potential breakage of the sha-256 generation upon new versions of npm.
- Uses FOD, so a change in sha-256 might get picked up much later, when someone tries to build on local.
Unresolved
- does the hash change depending on the platform ?

It seems to me that this needs package maintainers to be active in maintaining their package. let's not forget that the current update process is so intense, that people are getting discouraged from contributing altogether (reference in the HN post). Perhaps we could let package maintainers decide what they prefer ?

The discussion has surfaced that this could be a temporary solution waiting for a better one. I still don't see the problem of going forward with this and then changing/pushing for another solution when it's ready.

`npmlock2nix`:
Pros:
- does not use FOD or IFD
Cons:
- only usable with a package-lock.json, so potentially there would be a need to commit lockfiles

I don't see why this npmlock2nix cannot co-exist with fetchNodeModules. There are potentially problems left, but I feel if we mark this as beta in the documentation, we could potentially get more feedback from direct usage today. If somebody has a lock-file for which npmlock2nix is broken, they can try fetchNodeModules.

Regarding potentially different hashes on different platforms. One dirty solution could be to just have an if statement with the different platforms. It might not look pretty, but if with this 'trick' we can reduce the amount of generated code today, IMHO it's still a win.

regarding the potential bug of the current implementation
@zhaofengli I've tested the code you gave. The hash is different from me, this is what builds
```
nodePackages.fetchNodeModules {
  src = fetchFromGitHub {
    owner = "nttgin";
    repo = "BGPalerter";
    rev = "v1.28.1";
    sha256 = "sha256-Y0atkJfZxjUOGPQ3goXS/YD5SsX9ZjpbM0Nc5IuaFP4=";
  };

  makeTarball = false;
  exposeBin = true;
  sha256 = "sha256-UJuev/bqTTWjsrWiFWCSWZBlFxfaZH0X6IiQypbT+04=";
}
```
could you give more detail on the bug you are encountering ?
I've tried building this inside of nixpkgs, and I couldn't see anything out of the ordinary, inside the result I have
```
❯ ls
Permissions Size User Date Modified Name
dr-xr-xr-x     - root  1 Jan  1970  bin
dr-xr-xr-x     - root  1 Jan  1970  lib
```
(which node_modules inside lib)

I have time to test this more, I am happy to help here.

One more thing that has not been addressed is how this would play out with yarn.
Basically at the moment, a lot of people go against the upstream lock and commit a yarn.lock (even if upstream is using package-lock.json for example) just because yarn2nix is more reliable than node2nix. We could potentially get rid of the generated code there.
The problem would be left when upstream packages originally use yarn.lock.

After we have resolved some of the questions left, would it make sense to merge this and keep it in an alpha state ? What I mean by that is that we can test with several packages for a couple of weeks and not add anything in the documentation. It seems to me that the chances for people to find this if it's not documented are very slim. After we have covered enough test cases, we could add the documentation and mark it as beta.

please let me know if I've forgot anything.

06kellyjac commented:
Doesnt add much but theres also pnpm lock files.
Also there might be some nuance to capture here around yarn v2 aka berry which vendors modules.

happysalada commented:
I was thinking we could try creating a separate fetcher for yarn. I wanted to test the reproducibility of yarn.
At the moment there is a [Peertube PR](https://github.com/NixOS/nixpkgs/pull/119110) that adds 28k lines of yarn.nix, that would be my prime target for trying the fetcher.
I believe that PR is just waiting a blessing to add those lines.

zhaofengli commented:
@yu-re-ka @happysalada 

> ```
> nodePackages.fetchNodeModules {
>   src = fetchFromGitHub {
>     owner = "nttgin";
>     repo = "BGPalerter";
>     rev = "v1.28.1";
>     sha256 = "sha256-Y0atkJfZxjUOGPQ3goXS/YD5SsX9ZjpbM0Nc5IuaFP4=";
>   };
> 
>   makeTarball = false;
>   exposeBin = true;
>   sha256 = "sha256-UJuev/bqTTWjsrWiFWCSWZBlFxfaZH0X6IiQypbT+04=";
> }
> ```

Thank you for testing! Looks like we have some reproducibility issue happening here. Could you upload the NARs and tars (for `makeTarball = true`) along with the expressions that you have?

> could you give more detail on the bug you are encountering ?
> I've tried building this inside of nixpkgs, and I couldn't see anything out of the ordinary, inside the result I have

Curiously, I still can't manage to build with `makeTarball = false` and `exposeBin = true`. This is very, very weird:

```
$ nix-build test.nix
this derivation will be built:
  /nix/store/3k8fd2hrbynplnkjyq97a79mxmbnk6sk-source-node_modules.drv
building '/nix/store/3k8fd2hrbynplnkjyq97a79mxmbnk6sk-source-node_modules.drv'...
unpacking sources
unpacking source archive /nix/store/1cl0ph8bybbk8gpiaxwrrzkhp99vymhc-source
source root is source
patching sources
building
Running npm ci --no-update-notifier  --ignore-scripts --production
added 204 packages in 1.62sactTree: sill extract rxjs@6.6.7 extracted to /build/bm
installing
Exposing executables
- bunyan
- fast-file-logger
- js-yaml
- mime
- mkdirp
- ncp
- node-gyp-build
- node-gyp-build-optional
- node-gyp-build-test
- report-latency
- rimraf
- semver
- sshpk-conv
- sshpk-sign
- sshpk-verify
- uuid
/nix/store/7k43ld2z0jbiglhrw3c18s7y0vgqv6d7-source-node_modules
$ ls -lah result
lrwxrwxrwx 1 zhaofeng users 63 Sep 10 21:14 result -> /nix/store/7k43ld2z0jbiglhrw3c18s7y0vgqv6d7-source-node_modules
$ ls -lah result/
ls: cannot access 'result/': No such file or directory
$ ls /nix/store/7k43ld2z0jbiglhrw3c18s7y0vgqv6d7-source-node_modules
ls: cannot access '/nix/store/7k43ld2z0jbiglhrw3c18s7y0vgqv6d7-source-node_modules': No such file or directory
$ nix --version
nix (Nix) 2.4pre20210802_47e96bb
```

happysalada commented on pkgs/build-support/fetchnodemodules/default.nix:
```diff
@@ -0,0 +1,102 @@
+  name = name_;
+  nativeBuildInputs = [ nodejs_ makeWrapper ];
+
+  phases = "unpackPhase patchPhase buildPhase installPhase";
```
just noticed, probably not important at all, but the phases attribute is being (already?) deprecated.

ilkecan replied:
I wasn't aware of this. If anyone else wants to follow up on this, [this](https://github.com/NixOS/nixpkgs/issues/28910) looks like the related issue.

happysalada commented:
I've got some more informations on the mystery.
Actually I'm running into the same problem that you have. However it doesn't happen straight away for me. My computer being quite slow, just after creating the derivation, I can cd into it. I can't do anything else though. If I try `nix path-info` it says the path is not valid.
looking into the store though (`ls /nix/store | rg node_modules`) there is a valid path being created. Could it be possible that the nix-daemon is moving the original path to a new one ?
I have the same nix version as you do.
The resulting nar file is quite big (30 M) and I can only get it from the correct path in the nix-store (found with `rg`).
[nar.tar.gz](https://github.com/NixOS/nixpkgs/files/7153103/nar.tar.gz)

happysalada commented:
Here is the output of the makeTar
[node_modules_maketar.tar.gz](https://github.com/NixOS/nixpkgs/files/7153136/node_modules_maketar.tar.gz)

happysalada commented:
The weird thing in that derivation is that it does not get garbage collected and
```
❯ nix-store --query --roots /nix/store/mmds1ld337i6j4mkxsnlhcxbkgasmdml-source-node_modules
{censored} -> /nix/store/mmds1ld337i6j4mkxsnlhcxbkgasmdml-source-node_modules
```

happysalada commented:
I've tried giving the derivation a name to see if that would change anything (passing a `name` argument to fetchNodeModules`, the same thing happens, two paths are created, an invalid one and a valid one. The symlink is made on the invalid one).

happysalada commented on pkgs/build-support/fetchnodemodules/default.nix:
```diff
@@ -0,0 +1,102 @@
+    fi
+
+    export SOURCE_DATE_EPOCH=1
+    export npm_config_cache=/tmp
```
isn't it a problem to refer to temp directly ? (just wondering)
lots of other places have use `mktemp`

roberth replied:
In the darwin sandbox you can't use `/tmp` iirc. `mktemp` uses the right tmp from the environment variables and creates a unique directory, which may prevent a surprise one day.

happysalada commented on pkgs/build-support/fetchnodemodules/default.nix:
```diff
@@ -0,0 +1,102 @@
+
+  customArgs = [ "name" "hash" "sha256" ];
+in stdenvNoCC.mkDerivation ({
+  inherit src production runScripts makeTarball preferLocalBuild;
```
you forgot to inherit exposeBin, or am I missing something ?

happysalada replied:
saeme with the npmFlags ?

happysalada replied:
Last comment on this, I don't know what is the best practice here. But rather than inheriting variables that are unrelated to mkDerivation (I had to search the codebase to see if those variables where used elsewhere), could it make sense to use those variables directly where they are required? Something like
- remove the inherit runScripts
- inside the build phase have something like
```
''
if  ! ${runScripts} ; then
        NPM_FLAGS+=" --ignore-scripts"
fi
''
```

zhaofengli replied:
Yes, I should add it there for clarity, even though it's pulled in with the other arguments in the last line with `//`.

happysalada commented on pkgs/build-support/fetchnodemodules/default.nix:
```diff
@@ -0,0 +1,102 @@
+    fi
+
+    # Whether to install dev dependencies
+    if [[ -n $production ]]; then
```
What I know about bash is that the `-n` checks if a variable is non empty.
What is a little confusing to me is that production is either true or false, but it's never empty.
Perhaps this is a behavior of a mkDerivation that will convert variable it inherits as empty if they are false. It seems a littlle unintuitive to me. I just want to confirm here that this is the case.

happysalada commented on pkgs/build-support/fetchnodemodules/default.nix:
```diff
@@ -0,0 +1,102 @@
+  installPhase = ''
+    if [[ -n $makeTarball ]]; then
+        # Produce a deterministic tarball
+        tar -czf $out --owner=0 --group=0 --numeric-owner --format=gnu \
```
one thing that is curious to me is that cargo by default makes a tarball, there is no way to not make a tarball.
([ref](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/rust/fetchCargoTarball.nix))
I wonder if anybody what motivated that decision.

Checking the go code, it does not make a tarball at all. ([ref](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/go-modules/generic/default.nix#L64))

I wonder why
- go took an opposite approach from rust
- why it would be good to have both ?

I'm guessing the tarball can help if you need to remote-build ?

happysalada commented on pkgs/build-support/fetchnodemodules/default.nix:
```diff
@@ -0,0 +1,102 @@
+, hash ? ""
+, sha256 ? ""
+, name ? null
+, production ? true
```
perhaps the default should be false here ?
My impression is that fetchNodeModules will fetch the node_modules. There is therefore a high chance that people will use those dependencies to build something. Building will require dev dependencies. I don't see a context in which it would make sense to not have the dev dependencies in the node_modules (since you won't be able to build the related project with it). Actually if you want to optimize the closure size, you might want to first have a dev dependencies, build the project then replace the node_modules folder with only production dependencies.

dotlambda replied:
Yes, the documentation should encourage packagers to run `npm prune --production` after building.

happysalada replied:
ok, then it's a matter of changing what the production flag does. At the moment it only installs dependencies and not devDependencies.

happysalada commented on pkgs/build-support/fetchnodemodules/default.nix:
```diff
@@ -0,0 +1,102 @@
+        cp -r node_modules $out/lib
+
+        # Expose executables in node_modules/.bin
+        if [[ -n $exposeBin ]]; then
```
just wondering here what the use case for this would be.
Would we want to expose binaries from the node_modules of a project ?
Were you thinking of `webpack` and those build executables for example ?
I'm wondering if it's not better to provide the node_modules folder as is and let the consumers of that folder handle the binary paths ?
(this is the approach that node2nix takes if I remember correctly). basically if you want in your consuming derivation use a single binary you can address it by path, if you want to make all of those available, you can do `export PATH=node_modules/.bin`

zhaofengli replied:
(Sorry for the late response)

Yes, exactly. It's mostly there for convenience since some build scripts attempt to run those commands directly. I can do just fine without them, but we still need to resolve the other reproducibility issues.

happysalada commented:
In the interest of discussion I went ahead and did an implementation for yarn
https://github.com/NixOS/nixpkgs/pull/138233
(handles both yarn.lock and package-lock.json)
I would be happy to get feedback on it.
My thinking is that if it takes to get another 3 months before we get npmlock2nix in nixpkgs but we can get a fetcher in next week, then it's that much earlier that we can start making a transition for the js eco-system.

roberth commented:
Have we considered the old `haskellPackages`-like workflow?

 - contributors submit changes to the config file only
 - update automation runs on a branch `node-updates` (like `haskell-updates`)
 - hydra builds it and runs the relevant tests
 - merge weekly

This is the old workflow though. Nowadays contributors do generate the exprs, because it's fast, but `haskellPackages` has done without that for a long time.
For this to be feasible on node, contributors would have to run with a minimal config file which is sufficient for their package only, so they can regenerate the expressions and test in a reasonable amount of time.

It does avoid the technical issues.

zhaofengli commented:
To move this forward, I have removed the `exposeBin` functionality and rebased the PR. Additionally, the following changes were made:

- The value of `production` is embedded in the output path, to alleviate the problem caused by forgetting to change the hash in this simple case so that it's guaranteed to cause a visible failure.
- `package.json` and `package-lock.json` are now injected into the output, so the consumer can ensure that the packages match the source it's building. As @roberth has pointed out, changes like this will silently break existing users, so we must ensure that the default behavior can satisfy most usecases and be expected remain unchanged for the foreseeable future.

For the missing output problem, an issue for this should be created for Nix, once we figure out a minimal repro example.

---

I looked a bit more into the tarball reproducibility issue that @happysalada and @yu-re-ka mentioned, and it appears that the only differences come down to the permissions of the symlinks in the tar. For example:

```
│ │ -lrwxrwxrwx   0        0        0        0 1970-01-01 00:00:01.000000 node_modules/restify/node_modules/.bin/semver -> ../semver/bin/semver.js
│ │ +lrwxr-xr-x   0        0        0        0 1970-01-01 00:00:01.000000 node_modules/restify/node_modules/.bin/semver -> ../semver/bin/semver.js
```
(`-` is what I got, and `+` is @happysalada's tarball)

For reference, I'm using ZFS as my filesystem.

happysalada commented:
Thank you for your work on this!

I've tested again and here is the thing that works for me
```
{ nodePackages, fetchFromGitHub }: 

nodePackages.fetchNodeModules {
  name = "bgp-node-modules";
  src = fetchFromGitHub {
    owner = "nttgin";
    repo = "BGPalerter";
    rev = "v1.28.1";
    sha256 = "sha256-Y0atkJfZxjUOGPQ3goXS/YD5SsX9ZjpbM0Nc5IuaFP4=";
  };

  makeTarball = false;
  exposeBin = true;
  sha256 = "sha256-/x8XwHxIEYS3IxEuk1jp1wYyJIV11DdSXUU+DCkrQ1g=";
}
```

(just to confirm if we still have a different hash on different platforms)

The good news is that the bug of the disappearing path after building is gone with this new update.

I think the main point is to figure out if the hash is still different on different platforms. If it is, what to do about it.

There are a couple of additional nits
- using `mktemp` instead of `/tmp`
- perhaps defaulting production to false ? I was thinking that most people wanting to use node_modules are going to use those to build something and so including dev modules seems the default to me.


@roberth I like the idea. People could generate their own node-packages to test when adding a package. When they are confident it works, they just commit the result of the expression, and then weekly the version update can be run. There might still be some hiccups, but it could be a short term solution. I think it would just be a matter of documenting the workflow properly.

roberth commented:
@happysalada would you like to lead such an effort? The `haskellPackages` team has shown that a few committed members can have a great impact.

happysalada commented:
@roberth thank you for your idea. I am a little at capacity right now, but I have added testing this approach on my task list for as soon as I get a bit more time.

happysalada commented:
What is your level of interest on this ?
It needs to be tested with several packages to see if the hash difference appears on different platforms for different packages. If it does, I don't think it's a problem, it's just a matter of making nix expressions that take into account the different hashes for different platform. (I think the vscode package has similar things).

if you are interested in doing 3 tests with js packages you are interested in, and reporting the results, that would be amazing.
After that, we can try to make a decision whether to merge or not. I'm personally in favour.

zhaofengli commented:
I've been using this downstream, but we should push this forward. I'll rebase the PR in a bit, and add the `bgpalerter` package in as well so it can be more easily tested.

RaitoBezarius commented:
> What is your level of interest on this ? It needs to be tested with several packages to see if the hash difference appears on different platforms for different packages. If it does, I don't think it's a problem, it's just a matter of making nix expressions that take into account the different hashes for different platform. (I think the vscode package has similar things).
> 
> if you are interested in doing 3 tests with js packages you are interested in, and reporting the results, that would be amazing. After that, we can try to make a decision whether to merge or not. I'm personally in favour.

Between medium and high, I am not sure though I have different platforms alas (only `x86_64-linux`, but maybe I can ask for `aarch64-linux` community builder to test this platform), I can give it a try on some packages I know.

winterqt commented:
Hi all.

Over the past week or so, I've been working on a solution to this issue. (I actually didn't even know this PR was a thing before starting work on mine, until someone told me about it.)

It populates a cache directory to avoid the issue of different hashes occurring across platforms, and because of its superior reproducibility, is most likely the better solution to push forward.

There are more details in #189539 if you're interested in taking a look -- I welcome any and all comments/suggestions.

zhaofengli commented:
I haven't tested https://github.com/NixOS/nixpkgs/pull/189539 yet, but I think the approach there is less prone to breakages. Let's close this and focus our efforts on that.
